prboom2/src/gl_shader.c

```diff
@@ -53,217 +54,551 @@
+typedef union
+{
+  int i[2];
+  float f[2];
+} shader_uniform_value_t;
 
-  GLEXT_glUseProgramObjectARB(0);
+typedef struct
+{
+  shader_t* shader;
+  shader_uniform_value_t unifs[MAX_UNIFORMS];
+} shader_frame_t;
+
+static void glsl_ShaderSrcInit(shader_source_t* src)
+{
+  memset(src, 0, sizeof(*src));
 }
 
-static void get_fuzz_shader_bindings()
+static void glsl_ShaderSrcDestroy(shader_source_t* src)
 {
-  int idx;
+  if (src->strs)
+    Z_Free(src->strs);
+  if (src->lens)
+    Z_Free(src->lens);
+  glsl_ShaderSrcInit(src);
+}
 
-  fuzz_unifs.tex_d_index = GLEXT_glGetUniformLocationARB(sh_fuzz->hShader, "tex_d");
-  fuzz_unifs.ratio_index = GLEXT_glGetUniformLocationARB(sh_fuzz->hShader, "ratio");
-  fuzz_unifs.seed_index = GLEXT_glGetUniformLocationARB(sh_fuzz->hShader, "seed");
+static void glsl_ShaderSrcAppend(shader_source_t* src, const GLchar* str, GLint len)
+{
+  unsigned int size = src->size;
 
-  GLEXT_glUseProgramObjectARB(sh_fuzz->hShader);
+  if (size && src->lens[size - 1] >= 0 && src->strs[size - 1] + src->lens[size - 1] == str)
+  {
+    // Fast path -- expand last string to include more of source buffer
+    src->lens[src->size - 1] += len;
+    return;
+  }
 
-  idx = GLEXT_glGetUniformLocationARB(sh_fuzz->hShader, "tex");
-  GLEXT_glUniform1iARB(idx, 0);
+  if (src->size == src->cap)
+  {
+    if (src->cap == 0)
+      src->cap = 8;
+    else
+      src->cap *= 2;
 
-  idx = GLEXT_glGetUniformLocationARB(sh_fuzz->hShader, "fuzz");
-  GLEXT_glUniform1iARB(idx, 1);
+    src->strs = Z_Realloc(src->strs, src->cap * sizeof(*src->strs));
+    src->lens = Z_Realloc(src->lens, src->cap * sizeof(*src->lens));
+  }
 
-  GLEXT_glUseProgramObjectARB(0);
+  src->strs[src->size] = str;
+  src->lens[src->size] = len;
+  ++src->size;
 }
 
-void glsl_Init(void)
+static void glsl_ShaderLookup(const char* name, GLchar const** text, GLint* len)
 {
-  sh_indexed = gld_LoadShader("glvp", "glfp_idx");
-  get_indexed_shader_bindings();
+  int lump = W_CheckNumForName2(name, ns_prboom);
+
+  if (lump == LUMP_NOT_FOUND)
+    I_Error("Could not find shader source: %s\n", name);
+
+  *text = W_LumpByNum(lump);
+  *len = W_LumpLength(lump);
+}
+
+#define CSLEN(x) (sizeof((x)) - 1)
 
-  sh_fuzz = gld_LoadShader("glvp", "glfp_fuzz");
-  get_fuzz_shader_bindings();
+static void glsl_ShaderSrcAppendDefine(shader_source_t* src,
+                                       const shader_define_t* def)
+{
+  static const char cdefine[] = "#define ";
+  static const char cspace[] = " ";
+  static const char cnl[] = "\n";
+
+  glsl_ShaderSrcAppend(src, cdefine, CSLEN(cdefine));
+  glsl_ShaderSrcAppend(src, def->name, strlen(def->name));
+  glsl_ShaderSrcAppend(src, cspace, CSLEN(cspace));
+  glsl_ShaderSrcAppend(src, def->value, strlen(def->value));
+  glsl_ShaderSrcAppend(src, cnl, CSLEN(cnl));
 }
 
-static int ReadLump(const char *filename, const char *lumpname, char **buffer)
+static void glsl_ShaderSrcProcess(shader_source_t* src, const GLchar* text,

```

Comment:
It occurs to me that we could run the preprocessor as part of the build step when constructing the wad so the lump is already processed, and there's nothing to do at run time. I would leave the current PR as it is and take that separately if we decide to do it. What do you think?

Reply:
If we invest in build-time shader processing, we'd want to run glslangValidator on it to validate the shaders, in which case we could use it to do the preprocessing (`glslangValidator -E`).  This would bloat the data going into `dsda-doom.wad`, if that matters.